File: 0871-minimum-number-of-refueling-stops/0871-minimum-number-of-refueling-stops.py
# TC- O(nlogn)
# SC- O(n)

class Solution:
    def minRefuelStops(self, target: int, startFuel: int, stations: List[List[int]]) -> int:
        max_heap = []
        stations.append((target, float('inf')))
        min_stops = 0

        curr_fuel = startFuel
        curr_pos = 0
        for loc, fuel in stations:
            curr_fuel -= loc-curr_pos
            while max_heap and curr_fuel<0:
                curr_fuel += -heapq.heappop(max_heap)
                min_stops += 1
            if curr_fuel<0:
                return -1
            heapq.heappush(max_heap, -fuel)
            curr_pos = loc
        return min_stops

File: 0871-minimum-number-of-refueling-stops/0871-minimum-number-of-refueling-stops.cpp
// SC- O(n)
// TC- O(nlogn)

class Solution {
public:
    int minRefuelStops(int target, int startFuel, vector<vector<int>>& stations) {
        int currpos = 0;
        int currfuel = startFuel;
        stations.push_back({target, 0});
        priority_queue<int> max_heap;

        int ans = 0;

        for(int i=0; i<stations.size(); i++){
            currfuel = currfuel - (stations[i][0]-currpos);
            currpos = stations[i][0];

            while(currfuel<0){
                if(max_heap.empty())
                    return -1;
                currfuel += max_heap.top();
                max_heap.pop();
                ans++;
            }
            max_heap.push(stations[i][1]);
        }
        return ans;
    }
};